Scripts for DAHITI Data Downloading and Processing

Tjmahir20 opened the pull request:
The DAHITI_io holds some functions to download and process virtual stations data records from DAHITI. The example notebook depicts an use case of these functions for two different study regions.

Commit:
Scripts DAHITI Data Downloading and Processing

The DAHITI_io holds some functions to download and process virtual stations data records from DAHITI. The example notebook depicts an use case of these functions for two different study regions.

mn5hk commented on Example_Notebooks/Script_DAHITI_io.ipynb:
```diff
@@ -0,0 +1,992 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import dahiti_io\n",
```
Hi Mahir, great work.

I think perhaps something like
```python
import pyaltim
from pyaltim.io import dahiti_io
```

would be a better way to import `dahiti_io`. In my python environment, `import dahiti_io` was not working. While the line may work on your environment because of your folder structure, a standard approach would be to access the code from the `pyaltim` module as shown above.

mn5hk commented on Example_Notebooks/Script_DAHITI_io.ipynb:
```diff
@@ -0,0 +1,992 @@
+    }
+   ],
+   "source": [
+    "dahiti_io.download_DAHITI_stations(\"./Lake_Tana/\",\"Lake_Tana.geojson\")"
```
Hi Mahir, I currently do not have access to the geojson files. As such, I could not move ahead beyond the line in the review process.

Would it be possible to make some geojson files available within the github file for the example notebook? Overall, great work and happy coding :)

mn5hk reviewed (commented):
Hi Mahir, the code seems to be working overall. I've left couple minor reviews inline.

strawpants commented:
Hi Mahir,

I agree with Amin, if the json files are small, you can add them in a `Example_Notebooks/data` subfolder.

Later we may actually move the notebooks to `docs/source/notebooks`, so that they can be parsed and build into the html documentation (see for example the structure in [shxarray](https://github.com/ITC-Water-Resources/shxarray/tree/main/docs/) ). 

To prepare for a automatic document generation, it's best to document your `dahiti_io` functions [using numpy style docstrings]( https://www.sphinx-doc.org/en/master/usage/extensions/example_numpy.html), which helps a lot in automation of the document generation.

also (and I didn't set the right example myself), it may be simpler to just call the file `dahiti.py` (and I should change `hydroweb_io.py` to `hydroweb.py`), because `import pyaltim.io.dahiti` is already self-explanatory.

All the best,

Roelof

Commit:
Adding Data folder and Renaming the Dahiti_io  to Dahiti

- Made the geojsons available in the Example_Notebooks>Data folder.

- Changed Dahiti_io.py to Dahiti.py

Tjmahir20 commented:
Hi Amin and Roelof,
Thank you for all the feed backs.
I have uploaded the geojson files to the Data folder and renamed the script from _Dahiti_io_ to _Dahiti_.
I am trying to figure out the numpy-style docstrings and implement them in my scripts.
Thank you.

mn5hk commented:
Hi Mahir,

The updated code seems to be working well for the most part for me. Please note that in some parts, `dahiti_io` needs to be renamed to `dahiti` to reflect the change to the sub-module name.

Lastly, I received an error in the following line:
``` python
DF = dahiti.update_metadata_DAHITI("./Example_Notebooks/GBM_Basin/")
```

Perhaps you can diagnose the error (either in the `Script_DAHITI_io.ipynb` or in the `dahiti.py` `update_metadata_DAHITI` function.

Enlisting the error here:
``` python
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
/tmp/ipykernel_9832/81363063.py in ?()
----> 1 DF = dahiti.update_metadata_DAHITI("./Example_Notebooks/GBM_Basin/")

~/softwares/pyaltim/src/pyaltim/io/dahiti.py in ?(folder_path)
    149   df1 = pd.DataFrame(min_height_data)
    150   df2 = pd.DataFrame(max_height_data)
    151   df3 = pd.DataFrame(date_range)
    152 
--> 153   MetaDataFrame_DAHITI = pd.merge(df1, df2, on='Dahiti_ID', how='left')
    154   MetaDataFrame_DAHITI = pd.merge(MetaDataFrame_DAHITI, df3, on='Dahiti_ID', how='left')
    155 
    156   MetaDataFrame_DAHITI["Min_Height"] = MetaDataFrame_DAHITI["water_level_x"]

~/venvs/pyaltim/lib/python3.10/site-packages/pandas/core/reshape/merge.py in ?(left, right, how, on, left_on, right_on, left_index, right_index, sort, suffixes, copy, indicator, validate)
    166             validate=validate,
    167             copy=copy,
    168         )
    169     else:
--> 170         op = _MergeOperation(
    171             left_df,
    172             right_df,
    173             how=how,
...
   1912 
   1913         # Check for duplicates
   1914         if values.ndim > 1:

KeyError: 'Dahiti_ID'
Output is truncated. View as a [scrollable element](command:cellOutput.enableScrolling?f2c89b8b-0e79-4453-b6ad-db2982ea3435) or open in a [text editor](command:workbench.action.openLargeOutput?f2c89b8b-0e79-4453-b6ad-db2982ea3435). Adjust cell output [settings](command:workbench.action.openSettings?%5B%22%40tag%3AnotebookOutputLayout%22%5D)...


```

strawpants committed:
Rewrote connectors for dahiti and hydroweb, and added geoslurp entrypoints

strawpants committed:
fixed merge conflict

strawpants commented:
Dear Mahir and Amin,

I've made a fair amount of changes and reorganization of the files. The notebook demos are still missing the hydroweb case, and the current notebooks have some errors because I ran out of API calls. But I'll add an example hydroweb notebook soon.

strawpants committed:
add hydroweb example notebook

strawpants commented:
OK. I've now also added a hydroweb example notebook. I'm merging this into main
